### PR TITLE
Bit Map Changes

### PR DESCRIPTION
**Pull request by pranavantwso2**
Added SubscriptionBitMapHandler class. And modified SubscriptionStore class.
Check and Review.

### CODE REVIEW

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 1]
```diff
@@ -0,0 +1,671 @@
...
+                    bitMapClustered.add(newBitMapForithConstituent);
+                }
+
+                /** Set the bit for the subscription constituent part
+                 *
+                 */
+                Map<String, BitSet> bitMapForIthConstituent = bitMapClustered.get(i);
+
+                if ("#".equals(destinations[i]) || "*".equals(destinations[i])) {
+                    for (Map.Entry<String, BitSet> mapEntryForIthBitMap : bitMapClustered.get(i).entrySet())
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Will we get # and * to constants?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 2]
```diff
@@ -0,0 +1,671 @@
...
+import java.util.BitSet;
+
+/**
+ * This class is responsible for adding of the subscriptions into the bitmap, matching and the removal
+ * of subscriptions from bitmap
+ */
+public class SubscriptionBitMapHandler {
+
+    private static final String CONSTITUENT_TOPIC_CONSTANT = "other";
+    private final String SPECIAL_CHARACTER_FOR_NULL = "*";
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Explain the usage!

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 3]
```diff
@@ -0,0 +1,671 @@
...
+                        } else {
+                            if ("#".equals(destinations[constituentPart]))
+                                mapEntryForithBitMap.getValue().set(columnIndexOfTheSubscriptionInBitMap);
+                        }
+                    }
+                } else {
+                    BitSet bitSetForspecificConstituentPartOfSubscription;
+                    bitSetForspecificConstituentPartOfSubscription = bitMapForithConstituent.get(destinations[constituentPart]);
+                    if (bitSetForspecificConstituentPartOfSubscription == null)
+                        bitSetForspecificConstituentPartOfSubscription = (BitSet) bitMapLocal.get(constituentPart).get("other")
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: a magical static string "other". What about defining a constant? and comment on specialty of "other"?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java (hunk 1 above)
Comment (pranavantwso2) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Yes. We will get that.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java (hunk 3 above)
Comment (pranavantwso2) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Fixed.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java (hunk 2 above)
Comment (pranavantwso2) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Explained.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+                 */
+                if (bitMapLocal.size() <= constituentPart) {
+                    Map<String, BitSet> newBitMapForIthConstituent;
+                    newBitMapForIthConstituent = Collections.synchronizedMap(new HashMap<String, BitSet>());
+
+                    BitSet bitSet;
+                    bitSet = new BitSet(localSubscriptionCount);
+                    bitSet.set(0, localSubscriptionCount);
+
+                    for (int deleted : deletedLocals) bitSet.clear(deleted);
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: shouldn't we be using compound statements ? better maintainability.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 5]
```diff
@@ -0,0 +1,682 @@
...
+            // Take the entry for the constituent part
+            if (null != bitMapOfIthConstituent.get(destinations[constituentCount]))
+                bitSetOfIthConstituent = (BitSet) bitMapOfIthConstituent.get(destinations[constituentCount]).clone();
+            else
+                bitSetOfIthConstituent = (BitSet) bitMapOfIthConstituent.get(CONSTITUENT_TOPIC_CONSTANT).clone();
+
+            if (constituentCount != 0) {
+                results.and(bitSetOfIthConstituent);
+            } else
+                results = bitSetOfIthConstituent;
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Compound statement please.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java (hunk 4 above)
Comment (pranavantwso2) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Fixed the issue.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java (hunk 5 above)
Comment (pranavantwso2) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Fixed the issue.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 6]
```diff
@@ -0,0 +1,688 @@
...
+    public SubscriptionBitMapHandler() {
+        localSubscriptions = new ConcurrentHashMap<Integer, Map<String, LocalSubscription>>();
+        clusteredSubscriptions = new ConcurrentHashMap<Integer, Map<String, AndesSubscription>>();
+        localSubscriptionMapping = new ConcurrentHashMap<String, Integer>();
+        clusteredSubscriptionMapping = new ConcurrentHashMap<String, Integer>();
+        bitMapLocal = new ArrayList<Map<String, BitSet>>();
+        bitMapClustered = new ArrayList<Map<String, BitSet>>();
+        localSubscriptionCount = 0;
+        clusteredSubscriptionCount = 0;
+        deletedClusters = new ArrayList<Integer>();
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: rename to deletedClusterSubscriptions maybe ?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 7]
```diff
@@ -0,0 +1,688 @@
...
+            //Loop through the constituent parts
+            for (int i = 0; i < destinations.length; i++) {
+                /**
+                 * When the constituent table is not created before
+                 * create new one Ex - If the topic is WSO2.Products.MB
+                 * But before we only had the subscriptions and topics with maximum of
+                 * 2 constituentsAdd a new table for the constituent part 3
+                 */
+                if (bitMapClustered.size() <= i) {
+                    Map<String, BitSet> newBitMapForithConstituent;
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Can we use a cleaner name ? e.g. bitMapForCurrentConstituent ? (will help to someone else going through the code)

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 8]
```diff
@@ -0,0 +1,688 @@
...
+
+                /** Set the bit for the subscription constituent part
+                 *
+                 */
+                Map<String, BitSet> bitMapForIthConstituent = bitMapClustered.get(i);
+
+                if ("#".equals(destinations[i]) || "*".equals(destinations[i])) {
+                    for (Map.Entry<String, BitSet> mapEntryForIthBitMap : bitMapClustered.get(i).entrySet())
+                        if (!SPECIAL_CHARACTER_FOR_NULL.equals(mapEntryForIthBitMap.getKey())) {
+                            mapEntryForIthBitMap.getValue().set(columnIndexOfTheSubscriptionInBitMap);
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Can we have a cleaner variable name ?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 9]
```diff
@@ -0,0 +1,688 @@
...
+     * @param isLocal     indicates local or andes subscription
+     */
+    private void removeSubscription(int toBeRemoved, boolean isLocal) {
+        ArrayList<Map<String, BitSet>> bitMap = isLocal ? bitMapLocal
+                : bitMapClustered;
+
+        // clear the bit related with the deletion of subscription
+        for (int i = 0; i < bitMap.size(); i++) {
+            Map<String, BitSet> bitSetMap = bitMap.get(i);
+            for (Map.Entry<String, BitSet> entryMap : bitSetMap.entrySet()) entryMap.getValue().clear(toBeRemoved);
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Code formatting needed.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 10]
```diff
@@ -0,0 +1,688 @@
...
+    /**
+     * To get the matching subscriptions for a destination
+     *
+     * @param destination routing pattern of the message
+     */
+    public ArrayList<LocalSubscription> findMatchingLocalSubscriptions(String destination) {
+        /**
+         * Split the routing key of the subscriptions into the constituent parts
+         */
+        String[] destinations = destination.split("\\.");
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Can move to a constant ?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 11]
```diff
@@ -0,0 +1,688 @@
...
+         */
+        for (int constituentCount = 0; constituentCount < bitMapLocal.size() && constituentCount < destinations.length; constituentCount++) {
+
+            Map<String, BitSet> bitMapOfIthConstituent = bitMapLocal.get(constituentCount);
+            BitSet bitSetOfIthConstituent;
+
+
+            // Take the entry for the constituent part
+            if (null != bitMapOfIthConstituent.get(destinations[constituentCount])) {
+                bitSetOfIthConstituent = (BitSet) bitMapOfIthConstituent.get(destinations[constituentCount]).clone();
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Use a cleaner variable name

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 12]
```diff
@@ -0,0 +1,688 @@
...
+                results.and(bitSetForNull);
+
+            }
+
+        }
+
+        for (int numberOfLocalSubscriptions = 0; numberOfLocalSubscriptions < localSubscriptionCount; ) {
+            int nextMatchingLocalSubIndex = results.nextSetBit(numberOfLocalSubscriptions);
+            if (nextMatchingLocalSubIndex >= 0) {
+                for (LocalSubscription l : localSubscriptions.get(nextMatchingLocalSubIndex).values()) {
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Use "localSubscription" as a variable name ?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 13]
```diff
@@ -0,0 +1,688 @@
...
+                        .get(destinations.length);
+                BitSet bitSetForOther = map.get(SPECIAL_CHARACTER_FOR_NULL);
+
+                results.and(bitSetForOther);
+
+            }
+        }
+
+        for (int i = 0; i < clusteredSubscriptionCount; ) {
+            int k = results.nextSetBit(i);
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: Use a meaningful variable name

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java [hunk 14]
```diff
@@ -0,0 +1,688 @@
...
+        return localSubscriptions.values();
+    }
+
+    /**
+     * Method to replace the all the previous subscriptions for a destination with a new list
+     * @param destination destination to be replaced
+     * @param newSubscriptionList new list of subscriptions
+     * @return the old list of the subscriptions
+     */
+    public List<AndesSubscription> getAllClustered(String destination, List<AndesSubscription> newSubscriptionList) {
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java]: rename to getAllClusteredSubscriptions ?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 15]
```diff
@@ -46,25 +46,39 @@
...
-        QUEUE_SUBSCRIPTION,
-        TOPIC_SUBSCRIPTION,
-        ALL
-    }
+    private boolean isBitmap;
+    /**
+     * If 0 native method
+     * If 1 bitmap
+     */
+    private TopicMathcingSelection topicMathcingSelection;
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: Check spelling

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 16]
```diff
@@ -77,10 +91,18 @@ public SubscriptionStore() throws AndesException {
      */
     public List<AndesSubscription> getAllSubscribersForDestination(String destination, boolean isTopic) throws AndesException {
         // returing empty arraylist if requested map is empty
-        if (isTopic) {
-            return new ArrayList<AndesSubscription>(clusterTopicSubscriptionMap.get(destination) == null ? new ArrayList<AndesSubscription>() : clusterTopicSubscriptionMap.get(destination));
+        if (isBitmap) {
+            if (isTopic) {
+                return new ArrayList<AndesSubscription>(subscriptionBitMapHandler.getAllClusteredSubscribedForDestination(destination) == null ? new ArrayList<AndesSubscription>() : subscriptionBitMapHandler.getAllClusteredSubscribedForDestination(destination));
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: make sure code formatting is applied

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 17]
```diff
@@ -109,17 +135,13 @@ public SubscriptionStore() throws AndesException {
...
             // hence we need to check both maps
-
-            // Topic map is only have topic messages
-            subscriptionList.addAll(getSubscriptionsInMap(destination,
-                    clusterTopicSubscriptionMap, SUBSCRIPTION_TYPE.ALL));
-
-            // Get durable topic subscriptions from Queue map
-            subscriptionList.addAll(getSubscriptionsInMap(destination,
-                    clusterQueueSubscriptionMap, SUBSCRIPTION_TYPE.TOPIC_SUBSCRIPTION));
+            if (isBitmap)
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: its better to use curly braces to define condition scopes all the time. (helps reading)

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 18]
```diff
@@ -612,12 +776,151 @@ public synchronized void createDisconnectOrRemoveClusterSubscription(AndesSubscr
...
+                if (log.isDebugEnabled())
+                    log.debug("DELETED Subscription from map. queue name:" + subscription.getTargetQueue() + ", Type: " + subscription.getTargetQueueBoundExchangeType());
+            }
+
+
+        }
+
+    }
 
     private void printSubscriptionMap(Map<String, List<AndesSubscription>> map) {
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: Add javadoc comments

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/TopicMathcingSelection.java [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.andes.subscription;
+
+public enum TopicMathcingSelection {
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/TopicMathcingSelection.java]: Add enum level comment.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 20]
```diff
@@ -741,11 +1087,6 @@ public synchronized void createDisconnectOrRemoveLocalSubscription(LocalSubscrip
...
         }
 
-        log.debug("\n\tUpdated local subscription maps\n");
-        this.printLocalSubscriptionMap(localQueueSubscriptionMap);
-        this.printLocalSubscriptionMap(localTopicSubscriptionMap);
-        log.debug("\n");
-
     }
 
     private LocalSubscription removeLocalSubscription(AndesSubscription subscription) throws AndesException {
```
Comment (hastef88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: javadoc comment ?

**Pull request merged**